application/comit_node/src/connection_pool.rs

```diff
@@ -0,0 +1,286 @@
+use crate::{
+    bam_api::rfc003::swap_config, node_id::NodeId, swap_protocols::rfc003::bob::BobSpawner,
+};
+use bam::{self, json};
+use futures::{
+    future::{self, Shared},
+    Future,
+};
+use std::{
+    collections::HashMap,
+    fmt::{self, Debug},
+    io,
+    net::SocketAddr,
+    sync::{Arc, Mutex, RwLock},
+};
+use tokio::{
+    self,
+    io::{AsyncRead, AsyncWrite},
+    net::TcpStream,
+};
+
+type BamClient =
+    Arc<Mutex<bam::client::Client<json::Frame, json::OutgoingRequest, json::Response>>>;
+
+enum ConnectionState {
+    Connected {
+        client: BamClient,
+    },
+    Connecting {
+        waiting_list: Shared<Box<dyn Future<Item = BamClient, Error = io::ErrorKind> + Send>>,
+    },
+    Disconnected,
+}
+
+impl Debug for ConnectionState {
+    fn fmt<'a>(&self, f: &mut fmt::Formatter<'a>) -> Result<(), fmt::Error> {
+        write!(f, "{}", &self)
+    }
+}
+
+impl fmt::Display for ConnectionState {
+    fn fmt<'a>(&self, f: &mut fmt::Formatter<'a>) -> Result<(), fmt::Error> {
+        use self::ConnectionState::*;
+        match self {
+            Connected { .. } => write!(f, "connected"),
+            Connecting { .. } => write!(f, "connecting"),
+            Disconnected => write!(f, "disconnected"),
+        }
+    }
+}
+
+/// Keeps track of connections per NodeId.
+/// For now we don't allow multiple connections from the same NodeId.
+#[derive(Default, Debug)]
+pub struct ConnectionPool {
+    connections: RwLock<HashMap<NodeId, Arc<Mutex<ConnectionState>>>>,
+}
+
+impl ConnectionPool {
```

Comment:
:wrench: Can we have 3 functions on `ConnectionPool` for 

1. establishing a new connection
2. putting a caller on the wait_list
3. retrieve the existing client

This would simplify the `client_for` method a bit and only concern it with delegating based on the current state.

For consistency, I would make each method log at the beginning what they are about to do. Currently, adding a caller to the wait_list is missing this log statement.